### PR TITLE
Adding Sprite and SpriteAnimation easier ways of loading

### PR DESCRIPTION
**Pull request by erickzanardo**
# Description

Proposal for a new simpler way of loading Sprites and SpriteAnimations, this PR implements a new class for Sprites/SpritesAnimation parameters and so we can have a nice shortcut method on the game class.

Once the proposal is discussed and accepted, I will update examples and docs.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

If something is unclear, please submit the PR anyways and ask about what you thought was unclear.

- [x] This branch is based on `v1.0.0`
- [x] This PR is targeted to merge into `v1.0.0` (not `master` or `develop`)
- [x] I have added an entry under `[next]` in `CHANGELOG.md`
- [x] I have formatted my code with `flutter format`
- [x] I have made corresponding changes to the documentation
- [x] I have added examples for new features in `doc/examples`
- [x] The continuous integration (CI) is passing

### CODE REVIEW

File: lib/sprite.dart [hunk 1]
```diff
@@ -4,18 +4,22 @@ import 'extensions/offset.dart';
 import 'extensions/vector2.dart';
 import 'palette.dart';
 
+class SpriteOpts {
```
Comment (spydon) [inline lib/sprite.dart:7]: Is this break out really necessary when there are only two values stored on here?
The other one makes sense since it has so many.

File: lib/sprite_animation.dart [hunk 2]
```diff
@@ -60,65 +94,30 @@ class SpriteAnimation {
     frames = sprites.map((s) => SpriteAnimationFrame(s, stepTime)).toList();
   }
 
-  /// Automatically creates a sequenced animation, that is, an animation based on a sprite sheet.
```
Comment (spydon) [inline lib/sprite_animation.dart:63]: Maybe this comment describing the arguments should be on the SpriteAnimationOpts.sequenced now?

**Review by spydon — COMMENTED** [review]
Looks good! Just unsure of the need for `SpriteOpts`.

File: lib/sprite.dart (hunk 1 above)
Comment (erickzanardo) [reply lib/sprite.dart:7]: I don't have strong opinions on that, I decided to use a new class to create a "standard" and be somehow similar to the SpriteAnimation, but both would work for me.

File: lib/sprite_animation.dart (hunk 2 above)
Comment (erickzanardo) [reply lib/sprite_animation.dart:63]: Yep. I have deleted it because it was quite outdated, when I do the docs, I will try to update and create new ones on `SpriteAnimationOpts`

File: lib/sprite.dart (hunk 1 above)
Comment (spydon) [reply lib/sprite.dart:7]: I think I still prefer keeping few arguments out of "opts"-classes, it's easy to end up in that pattern everywhere otherwise and too much of it would probably be considered an anti-pattern.

File: lib/sprite_animation.dart [hunk 3]
```diff
@@ -4,6 +4,40 @@ import 'dart:ui';
 import 'extensions/vector2.dart';
 import 'sprite.dart';
 
+class SpriteAnimationOpts {
+  int amount;
+  int amountPerRow;
+  Vector2 texturePosition;
+  Vector2 textureSize;
+  List<double> stepTimes;
```
Comment (luanpotter) [inline lib/sprite_animation.dart:12]: I think this should be a List<Frame> which has all this info. that is how we store it in a unified format.
then we can have a myriad of constructors for this class. for example, sequenced takes amount and a start position and follow as rows. we can rename sequenced to row and have another one for following a column. we can have one that figures everything out, and one that takes the List<Frame> as is.

File: lib/sprite.dart (hunk 1 above)
Comment (luanpotter) [reply lib/sprite.dart:7]: I think I prefer keeping this to Animation as well, Sprite is sufficiently simple and there is only one way to construct it (see my comment about animationOpt)

File: lib/sprite_animation.dart [hunk 4]
```diff
@@ -4,6 +4,40 @@ import 'dart:ui';
 import 'extensions/vector2.dart';
 import 'sprite.dart';
 
+class SpriteAnimationOpts {
```
Comment (luanpotter) [inline lib/sprite_animation.dart:7]: I am not sure `SpriteAnimationOpts` is the best name but I am not convinced either way. Maybe SpriteAnimationDef`? I feel it's more than options, it's the whole deal. Or SpriteAnimationData, or Frames or FramesDef

File: lib/sprite_animation.dart (hunk 3 above)
Comment (erickzanardo) [reply lib/sprite_animation.dart:12]: I didn't understood

File: lib/game/game.dart [hunk 5]
```diff
@@ -87,6 +89,19 @@ abstract class Game {
...
+    final image = await images.load(path);
+    return Sprite(image, opts);
+  }
+
+  /// Utility method to load and cache the image for a sprite animation based on its options
+  Future<SpriteAnimation> loadSpriteAnimation(
+      String path, SpriteAnimationOpts opts) async {
+    final image = await images.load(path);
+    return SpriteAnimation.fromImage(image, opts);
+  }
```
Comment (renancaraujo) [inline lib/game/game.dart:103]: Perhaps we could have this method in `Images` rather than here since there is nothing here that requires games properties?

Comment (erickzanardo) [reply lib/game/game.dart:103]: it requires the `images` property from game. I think it is better placed here since `Sprite` and `SpriteAnimations` are things that uses an `Image`. And I think it is nice to "center" this on the game, because we can add extension methods from our underlying packages, like a `loadSvg` from `flame_svg`.

File: lib/sprite.dart (hunk 1 above)
Comment (renancaraujo) [reply lib/sprite.dart:7]: opt classes are useful for:
- when options can live by itself and have inter-instance relations like `TextStyle` from flutter has `.apply` and `.compareTo` etc
- when we don't want to impact diferent places with API changes, ex:  adding or removing options here would not require to refactor on this `Sprite`contructor and `loadSprite` in the previous file. 
- users could extend it and change behavior, like `Decoration` form flutter

File: lib/sprite_animation.dart (hunk 4 above)
Comment (renancaraujo) [reply lib/sprite_animation.dart:7]: Whatever the name we pick, I don't think we should use acronyms. `SpriteAnimationOptions` LGTM.

**Review by renancaraujo — COMMENTED** [review]
I like the "options" idea and the usage of it. Left some opinions there.

File: lib/game/game.dart (hunk 5 above)
Comment (renancaraujo) [reply lib/game/game.dart:103]: > it requires the images property from game

Let me rephrase myself: it does require anything from game that wouldn't be accessible in a `Images` instance. 

>  "center" this on the game, because we can add extension methods from our underlying packages, like a loadSvg from flame_svg

M only concern here is not to end up with a sueprclass.

File: lib/sprite_animation.dart (hunk 4 above)
Comment (luanpotter) [reply lib/sprite_animation.dart:7]: you mean abbreviations? :p

File: lib/sprite_animation.dart (hunk 3 above)
Comment (luanpotter) [reply lib/sprite_animation.dart:12]: yes. I mean we can have a unified data type to represent the frames. it can contain frames of any size, any position, with any stepTime. So it will be a List<FrameData> where FrameData := { pos, size, stepSize }. Once we have that we can have a plethora of named constructors, like sequenced that takes just one pos, one size + amount and generates the frame data.

That will allow us to add a lot of constructors. I am proposing renaming sequenced to row and create a col one, for example. We can also have other ones.

Comment (erickzanardo) [reply lib/sprite_animation.dart:12]: Alright, I think I understood, to be honest, I don't think that it will be better, from a user view. But I will try thinking on something and make some tests and we can discuss later

File: lib/sprite.dart (hunk 1 above)
Comment (spydon) [reply lib/sprite.dart:7]: Since the arguments here are optional I don't see how those points are valid?

File: lib/sprite_animation.dart (hunk 3 above)
Comment (erickzanardo) [reply lib/sprite_animation.dart]: @luanpotter I have made the changes, it is way better now! Thanks for the suggestions, please let me know of what you think of that approach.

File: lib/game/game.dart [hunk 6]
```diff
@@ -87,6 +89,23 @@ abstract class Game {
...
+    Vector2 srcSize,
+    Vector2 srcPosition,
+  }) async {
+    final image = await images.load(path);
+    return Sprite(image, srcSize: srcSize, srcPosition: srcPosition);
+  }
+
+  /// Utility method to load and cache the image for a sprite animation based on its options
+  Future<SpriteAnimation> loadSpriteAnimation(
+      String path, SpriteAnimationData data) async {
```
Comment (luanpotter) [inline lib/game/game.dart:104]: trailing comma

File: lib/sprite_animation.dart [hunk 7]
```diff
@@ -4,6 +4,71 @@ import 'dart:ui';
...
+    Vector2 texturePosition,
+    Vector2 textureSize,
+    bool loop = true,
+  }) : this.variable(
+          amount: amount,
+          amountPerRow: amountPerRow,
+          texturePosition: texturePosition,
+          textureSize: textureSize,
+          loop: loop,
+          stepTimes: stepTime != null ? List.filled(amount, stepTime) : null,
```
Comment (luanpotter) [inline lib/sprite_animation.dart:68]: stepTimes is required non-null on variable. so this doesn't make a lot of sense. Probably should be list.filled(amount, stepTime ?? _DEFAULT_STEP_TIME)

**Review by luanpotter — APPROVED** [review]
I love this, great cleanup! I think now all places that create sprite animations can take that class and we can add a lot of constructors in a single place w/o duplicating.

For example, as a following, I am considering renaming sequenced to rowSequenced or something like that, and then we can add columnSequenced.

File: lib/sprite_animation.dart (hunk 7 above)
Comment (erickzanardo) [reply lib/sprite_animation.dart:68]: I explicitly added null here because the other constructor validates this on the assert. It is a pitty that bridge constructors on dart can't have asserts as well.

I think I prefer this solution over a default stepTime, what do you think?

**Review by spydon — APPROVED** [review]
Just missing some added docs and examples, other than that I think it looks good!

File: lib/sprite_animation.dart (hunk 7 above)
Comment (luanpotter) [reply lib/sprite_animation.dart:68]: so my understand of this code is that if stepTime here is null it will pass null to the other constructor and throw a NPE. but I haven't deeply analyzed the code, if this works it's fine for me :)

File: doc/examples/animations/lib/main.dart [hunk 8]
```diff
@@ -68,12 +52,33 @@ class MyGame extends BaseGame with TapDetector {
     add(reversedAnimationComponent);
   }
 
+  void addAnimation(double x, double y) {
```
Comment (spydon) [inline doc/examples/animations/lib/main.dart:55]: Maybe take a `Vector2` as input here instead?

File: doc/images.md [hunk 9]
```diff
@@ -192,25 +196,24 @@ class MyGame extends Game {
...
-* textureY : y position on the original image to start (defaults to 0)
-* textureWidth : width of each frame (defaults to null, that is, full width of the sprite sheet)
-* textureHeight : height of each frame (defaults to null, that is, full height of the sprite sheet)
-* removeOnFinish : a bool indicating if this AnimationComponent should be removed when the animation has reached its end
-
-So, in our example, we are saying that we have 8 frames for our player animation, and they are displayed in a row. So if the player height is also 16 pixels, the sprite sheet is 128x16, containing 8 16x16 frames.
-
 This constructor makes creating an Animation very easy using sprite sheets.
 
+In which you pass the image instance, and the frame data, which contains some parameters which can be used to describe the animation. Check the documentation on the constructors available on `SpriteAnimationFrameData` class to see all the parameters.
```
Comment (spydon) [inline doc/images.md:215]: "In which" can't be used since the subject is in the previous paragraph, I would write "In the constructor you pass an image instance, ..."

Third "which" should be "that"
-> "which contains some parameters that can ..."

File: lib/components/sprite_animation_component.dart [hunk 10]
```diff
@@ -19,62 +22,33 @@ class SpriteAnimationComponent extends PositionComponent {
     super.size.setFrom(size);
   }
 
+  /// Creates a component with an empty animation which can be set later
   SpriteAnimationComponent.empty();
 
-  SpriteAnimationComponent.sequenced(
+  /// Creates a SpriteAnimationComponent from a [size], an [image] and [data], check [SpriteAnimationData] for more info on the available options.
```
Comment (spydon) [inline lib/components/sprite_animation_component.dart:28]: Full stop instead of comma after [data]

**Review by spydon — APPROVED** [review]
Just some minor grammar stuff, otherwise it looks good!

File: doc/images.md [hunk 11]
```diff
@@ -192,25 +196,24 @@ class MyGame extends Game {
 }
 ```
 
-A better alternative to generate a list of sprites is to use the `sequenced` constructor:
+A better alternative to generate a list of sprites is to use the `fromFrameData` constructor:
 
 ```dart
   const amountOfFrames = 8;
-  SpriteAnimation a = SpriteAnimation.sequenced(imageInstance, amountOfFrames, textureWidth: 16.0);
+  SpriteAnimation a = SpriteAnimation.fromFrameData(
```
Comment (luanpotter) [inline doc/images.md:203]: final? (since you fixed that elsewhere)

File: doc/images.md [hunk 12]
```diff
@@ -109,10 +109,14 @@ You could also specify the coordinates in the original image where the sprite is
 
 ```dart
     final image = await loadImage();
-    Sprite playerFrame = Sprite(image, x: 32.0, width: 16.0);
+    Sprite playerFrame = Sprite(
```
Comment (luanpotter) [inline doc/images.md:112]: final? (since you fixed that elsewhere)

File: lib/components/sprite_animation_component.dart [hunk 13]
```diff
@@ -19,62 +22,33 @@ class SpriteAnimationComponent extends PositionComponent {
...
   @override
-  bool get shouldRemove => removeOnFinish && animation.isLastFrame;
+  bool get shouldRemove => removeOnFinish && (animation?.isLastFrame ?? false);
 
   @mustCallSuper
   @override
   void render(Canvas canvas) {
     super.render(canvas);
-    animation.getSprite().render(
+    animation?.getSprite()?.render(
```
Comment (luanpotter) [inline lib/components/sprite_animation_component.dart:51]: how could this be null?

File: lib/sprite_animation.dart [hunk 14]
```diff
@@ -4,6 +4,81 @@ import 'dart:ui';
...
 
+class SpriteAnimationFrameData {
+  /// Coordinates of the sprite of this Frame
+  final Vector2 srcPosition;
+
+  /// Size of the sprite of this Frame
+  final Vector2 srcSize;
+
+  /// The duration to display it, in seconds.
+  double stepTime;
```
Comment (luanpotter) [inline lib/sprite_animation.dart:15]: why are the others final and not this?

File: lib/sprite_animation.dart [hunk 15]
```diff
@@ -4,6 +4,81 @@ import 'dart:ui';
...
+  })  : assert(amountPerRow == null || amount >= amountPerRow),
+        assert(stepTimes != null) {
+    amountPerRow ??= amount;
+    texturePosition ??= Vector2.zero();
+    frames = List<SpriteAnimationFrameData>(amount);
+    for (int i = 0; i < amount; i++) {
+      final position = Vector2(
+        texturePosition.x + (i % amountPerRow) * textureSize.x,
+        texturePosition.y + (i ~/ amountPerRow) * textureSize.y,
+      );
```
Comment (luanpotter) [inline lib/sprite_animation.dart:55]: you can extract `texturePosition` outside like `texturePosition + Vector2(\n...`

File: lib/components/sprite_animation_component.dart (hunk 13 above)
Comment (erickzanardo) [reply lib/components/sprite_animation_component.dart:51]: We have an constructor which permits us to initialize the component without any animation. `SpriteAnimationComponent.empty`. I think our code had a bug when that constructor was used, so I added the elvis operator here to prevent NPE and improved the docs a little on `empty` constructor.

File: lib/sprite_animation.dart (hunk 14 above)
Comment (erickzanardo) [reply lib/sprite_animation.dart:15]: not sure, I probably just copied this from the previous implementation, but I changed it to final.

File: lib/sprite_animation.dart (hunk 15 above)
Comment (erickzanardo) [reply lib/sprite_animation.dart:55]: Not sure about that, I think I prefer the current implementation.

**Review by renancaraujo — APPROVED** [review]
